src/ir/module-utils.h

```diff
@@ -417,23 +427,14 @@ collectSignatures(Module& wasm,
 
       void visitExpression(Expression* curr) {
         if (curr->is<RefNull>()) {
-          maybeNote(curr->type);
+          counts.maybeNote(curr->type);
         } else if (auto* call = curr->dynCast<CallIndirect>()) {
           counts[call->sig]++;
         } else if (Properties::isControlFlowStructure(curr)) {
-          maybeNote(curr->type);
+          counts.maybeNote(curr->type);
           if (curr->type.isTuple()) {
             // TODO: Allow control flow to have input types as well
-            counts[Signature(Type::none, curr->type)]++;
-          }
-        }
-      }
-
-      void maybeNote(Type type) {
-        if (type.isRef()) {
-          auto heapType = type.getHeapType();
-          if (heapType.isSignature()) {
-            counts[heapType.getSignature()]++;
+            counts.note(Signature(Type::none, curr->type));
```

Comment:
Shouldn't we do this here too?
```cpp
if (type.isTuple()) {
  for (auto t : type) {
    counts.maybeNote(t);
  }
}
```
This is what I meant by doing tuple handling in `maybeNote` too... but I might be mistaken or something.

Reply:
I think on line 470 we should handle it? The count we add here adds a signature whose `results` is what your loop would traverse, and we traverse those values there.

Reply:
Oh right, I forgot about that code 👍